### PR TITLE
feat: Add `ignoreOnInitialization` option to no-shadow rule

### LINKED ISSUE #12687: no-shadow: allow a declaration's body to reuse the declared variable's name
**What rule do you want to change?**
no-shadow

**Does this change cause the rule to produce more or fewer warnings?**
Fewer warnings

**How will the change be implemented? (New option, new default behavior, etc.)?**
New default behavior

**Please provide some example code that this change will affect:**

```js
const person = people.find(person => person.name === "John");
```

**What does the rule currently do for this code?**
It currently produces a warning, claiming that the right-hand side `person` shadows the left-hand side `person`.

**What will the rule do after it's changed?**
It will not produce warnings when the purportedly-shadowed variable has not been declared.

I believe the current behavior is incorrect, as the variable has not (semantically) been declared at the point where the warning is generated, and so it can not be shadowed.

**Are you willing to submit a pull request to implement this change?**
Yes, though I would need some pointers on how to get started.

### PR DESCRIPTION
**Pull request by boutahlilsoufiane**
#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).
- [X] The team has reached consensus on the changes proposed in this pull request. If not, I understand that the evaluation process will begin with this pull request and won't be merged until the team has reached consensus.

#### What is the purpose of this pull request? (put an "X" next to an item)

[X] Changes an existing rule

#### What changes did you make? (Give an overview)
Fix the issue here: #12687

Adding a new option `ignoreOnInitialization` to prevent reporting variables on initialization statements.
This is an example:

```
const person = people.find(person => person.name === "John");
```

####  Is there anything you'd like reviewers to focus on?
No.

### CODE REVIEW

File: CHANGELOG.md
Comment (nzakas) [inline CHANGELOG.md:2745]: The change log is auto generated, so we shouldn’t make edits. Can you revert these changes?

File: lib/rules/no-shadow.js
Comment (nzakas) [inline lib/rules/no-shadow.js:81]: I think the names aren’t quite right here. I think this is a search operation, so maybe the names should reflect that, like:

```suggestion
        function findSelfOrAncestor(node, match) {
```

File: lib/rules/no-shadow.js
Comment (nzakas) [inline lib/rules/no-shadow.js:76]:
```suggestion
Searches from the current node through its ancestry to find a matching node. 
```

File: lib/rules/no-shadow.js
Comment (nzakas) [inline lib/rules/no-shadow.js:96]:
```suggestion
        function isInitPatternNode(variable, shadowedVariable) {
```

File: lib/rules/no-shadow.js
Comment (nzakas) [inline lib/rules/no-shadow.js:106]: It’s not clear what initNode1 and initNode2 represent. Can we rename them to better represent what they mean?

**Review by nzakas — CHANGES_REQUESTED** [review]
Functionally looks good. I just left some comments on how to clarify what is happening.

**Comment by boutahlilsoufiane** [comment]
> Functionally looks good. I just left some comments on how to clarify what is happening.
Hi @nzakas, thank you so much for reviewing the pull request.

**Comment by boutahlilsoufiane** [comment]
Hi @nzakas, thank you so much for reviewing the PR.

**Review by nzakas — APPROVED** [review]
LGTM. Just want to give others the chance to review too. Thanks!

File: lib/rules/no-shadow.js
Comment (mdjermanovic) [inline lib/rules/no-shadow.js:264]: This shouldn't change behavior when [`hoist`](https://eslint.org/docs/rules/no-shadow#hoist) option is set to `"all"`, since `hoist: "all"` means that shadowing should be reported regardless of where the variables are defined.

```js
/* eslint no-shadow: ["error", { hoist: "all" }] */

let x = foo((x, y) => {}); // Reports only y, but not x. Both should be reported.

let y;
```

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

**Comment by boutahlilsoufiane** [comment]
Hi @mdjermanovic, thank you for the explanation. The requested changes are addressed.

**Comment by ota-meshi** [comment]
I'm not the maintainer but I was interested in this fix so I looked at your code.
Is this fix intended to suppress reporting if it cannot be overridden (If misused, `ReferenceError` will occur) ?
If so, I think the following source code may need to be reported.
The following code will no longer be reported with this PR fix, but I don't think users will get a `ReferenceError` if they run it with `const a = 42;` removed.

```js
const a = fn(()=>{
    class C {
        fn () {
            const a = 42
            return a
        }
    }
    return new C()
})
```

What do you think?

**Comment by boutahlilsoufiane** [comment]
@ota-meshi I think we should report the source code, I will update the pull request to include your example.

**Comment by nzakas** [comment]
@mdjermanovic can you check to see if your comments were addressed?

**Comment by mdjermanovic** [comment]
Comments regarding the behavior of this rule with `hoist: "all"` are addressed - it should report shadowing regardless of the positions of variable declarations.

My other concern is that it isn't clearly defined what we want to fix (only parameters of functions that are arguments of function calls that are variable initializers should be excluded from shadowing that variable?), and why it should be fixed.

The original example from #12687 is:

```js
const person = people.find(person => person.name === "John");
```

This is shadowing, because at the moment when the inner `person` is created in an inner scope, the outer `person` already exists in the outer scope even though it may not be initialized yet. Furthermore, the outer `person` can be already initialized - we don't know whether the `person => person.name === "John"` function will be called right away or stored to be used later - so this shadowing doesn't have to be less potentially dangerous than shadowing a variable that is declared in a declaration that appears before this one. It should be only user's preference whether or not this example should be reported.

Problem is that the `"hoist"` option doesn't unambiguously specify how this example should be interpreted.

> The `hoist` option has three settings:
> 
> * `functions` (by default) - reports shadowing before the outer functions are defined.
> * `all` - reports all shadowing before the outer variables/functions are defined.
> * `never` - never report shadowing before the outer variables/functions are defined.

Is this "shadowing _before_ the outer variable is defined" or not? If we say that variable definition ends where the initializer ends, then it is "before", so `hoist: "functions"` or `hoist: "never"` should not report this example, and we have a false positive. On the other hand, it can be argued that the inner `person` identifier appears after the outer `person` identifier so it isn't "before", and therefore this rule works as intended (and this change might be undesirable for some users).

My opinion is that isn't clear if https://github.com/eslint/eslint/issues/12687 is a bug or intended behavior, and that we should resolve it by adding a new option rather than changing the current behavior.

**Comment by boutahlilsoufiane** [comment]
@mdjermanovic , I think you are right, the rule is working as expected, the inner variable person is shadowing the outer variable person, but developers are confused by this example:

`const person = people.find(person => person.name === "John");`

I think we should close the issue [#12687](https://github.com/eslint/eslint/issues/12687) and this PR and add the ambiguous example in the documentation.

For the code source mentioned by @ota-meshi the rule doesn't report it, we need to create a new issue and I will fix it.

**Comment by mdjermanovic** [comment]
I'm not suggesting that we close https://github.com/eslint/eslint/issues/12687, but to consider adding an option for that behavior.

> For the code source mentioned by @ota-meshi the rule doesn't report it, we need to create a new issue and I will fix it.

It works in the current version of this rule:

```js
/* eslint no-shadow: error */

const a = fn(()=>{
    class C {
        fn () {
            const a = 42 // 'a' is already declared in the upper scope
            return a
        }
    }
    return new C()
})
```

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXNoYWRvdzogZXJyb3IgKi9cblxuY29uc3QgYSA9IGZuKCgpPT57XG4gICAgY2xhc3MgQyB7XG4gICAgICAgIGZuICgpIHtcbiAgICAgICAgICAgIGNvbnN0IGEgPSA0MiAvLyAnYScgaXMgYWxyZWFkeSBkZWNsYXJlZCBpbiB0aGUgdXBwZXIgc2NvcGVcbiAgICAgICAgICAgIHJldHVybiBhXG4gICAgICAgIH1cbiAgICB9XG4gICAgcmV0dXJuIG5ldyBDKClcbn0pIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==)

**Comment by boutahlilsoufiane** [comment]
@mdjermanovic The requested changes are addressed.

**Comment by boutahlilsoufiane** [comment]
Hi @mdjermanovic, thank you so much for your time reviewing the PR. The requested changes are addressed.

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

**Comment by boutahlilsoufiane** [comment]
Hi @mdjermanovic! Thank you so much for your time, I will check your comments.

**Comment by boutahlilsoufiane** [comment]
Hi @mdjermanovic! Thank you so much for the examples. The comments are addressed.

**Review by mdjermanovic — COMMENTED** [review]
The behavior seems inconsistent in the following example because declarations for `x` and `y` are basically same, but the rule reports error only on `y`:

```js
/* eslint no-shadow: [2, { ignoreOnInitialization: true }] */

const x = foo(() => {
    const bar = () => {
        return x => {}; // no error on `x`
    }
    return bar;
});

const y = foo(() => {
    function bar() {
        return y => {}; // error on `y`
    }
    return bar;
});
```

It also looks inconsistent to me in this example:

```js
/* eslint no-shadow: [2, { ignoreOnInitialization: true }] */

const x = foo(() => {
    return {
        bar(x) {} // no error on `x`
    };
});

const y = foo(() => {
    class C {
        bar(y) {} // error on `y`
    }
    return new C();
});
```

Maybe the option should allow shadowing in callbacks, but not in functions that are nested in callbacks? For example:

```js
/* eslint no-shadow: [2, { ignoreOnInitialization: true }] */

const x = foo(x => {}); // allowed

const y = foo(() => y => {}); // not allowed
```

Thoughts? The option assumes that callbacks will be called right away and that therefore shadowing doesn't occur before the outer variable is initialized, but maybe it shouldn't assume the same for nested functions.

**Comment by boutahlilsoufiane** [comment]
Hi @mdjermanovic! Your explanations are great, thank you so much.
The mentioned examples should be reported because they will shadow the outer variable. I will fix that the next weekend. Thank you again for your awesome explanations.

**Comment by boutahlilsoufiane** [comment]
Hi @mdjermanovic! Thank you so much for the review. The comments are addressed.

File: lib/rules/no-shadow.js
Comment (mdjermanovic) [inline lib/rules/no-shadow.js:129]: This could find an arrow function or function expression that is not nested between definitions, but surrounds them and therefore should not affect the logic of this option.

```js
/* eslint no-shadow: ["error", { ignoreOnInitialization: true }] */

const x = foo(x => x); // allowed, correct

() => {
    const y = foo(y => y); // not allowed, incorrect?
}

```

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

**Comment by boutahlilsoufiane** [comment]
Hi @mdjermanovic, thanks a lot for the review and for the great explanation. I will try to fix the issue.

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

**Comment by boutahlilsoufiane** [comment]
Hi @mdjermanovic, thanks a lot for your instructions and for your time, you helped me a lot on this, I appreciate your time. ​​The comments are addressed.

File: lib/rules/no-shadow.js
Comment (mdjermanovic) [inline lib/rules/no-shadow.js:158]: This could find a common ancestor that isn't the one that declares the shadowed variable.

```js
/* eslint no-shadow: [2, { ignoreOnInitialization: true }] */

const foo = () => {
    let x;
    bar(x => x); // not reported, but it should be?
}

```

Maybe we could borrow the code that checks if a node is in variable's initializer from the `no-use-before-define` rule:

https://github.com/eslint/eslint/blob/355b23d0c4e050be4e53292f552a47c10ec6e00e/lib/rules/no-use-before-define.js#L190-L212

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

**Comment by boutahlilsoufiane** [comment]
Hi @mdjermanovic, thanks a lot for the snippet. The comments are addressed.

**Review by mdjermanovic — CHANGES_REQUESTED** [review]
@boutahlilsoufiane thanks for addressing all reviews! I think the most complex part of this change is now well done. I left some suggestions.

File: lib/rules/no-shadow.js
Comment (boutahlilsoufiane) [reply lib/rules/no-shadow.js:145]: We added this to treat the case of nested functions, this is an example:

`const x = foo(() => { const bar = () => { return x => {}; }; return bar; })`

I think we don’t need it anymore because now we check whether or not the outer scope of the shadowing variable equals the scope of the shadowed variable.

**Comment by boutahlilsoufiane** [comment]
Hi @mdjermanovic, thanks a lot for your contributions, you played a big role in this.

File: lib/rules/no-shadow.js
Comment (mdjermanovic) [inline lib/rules/no-shadow.js:84]:
```suggestion
         * @param {ASTNode} node a node to get.
         * @param {(node: ASTNode) => boolean} match a callback that checks whether or not the node verifies its condition or not.
         * @returns {ASTNode|null} the matching node.
```

As there's no type definition for `MatchCallback`, it might be better to just inline the expected signature here.

Also, by the implementation of this function and the way we're using it, it can return `null` (Program node's `parent` property value, when there's no matching ancestor).

File: lib/rules/no-shadow.js
Comment (mdjermanovic) [inline lib/rules/no-shadow.js:141]:
```suggestion
            let node = outerDef.name;
```

Since we now have an early exit when there's no outer def, we don't need to check that again here.

File: lib/rules/no-shadow.js
Comment (mdjermanovic) [inline lib/rules/no-shadow.js:143]:
```suggestion
            const location = callExpression.range[1];
```

We should check if the call expression is in the initializer. Otherwise, an unrelated call expression that surrounds both variables could affect the logic of this option.

```js
/* eslint no-shadow: ["error", { ignoreOnInitialization: true }] */

foo(
    () => {
        const x = x => x; // false negative
    }
);

```

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

File: docs/rules/no-shadow.md
Comment (mdjermanovic) [inline docs/rules/no-shadow.md:171]:
```suggestion
The `ignoreOnInitialization` option is `false` by default. If it is `true`, it prevents reporting shadowing of variables in their initializers when the shadowed variable is presumably still uninitialized.

The shadowed variable must be on the left side. The shadowing variable must be on the right side and declared in a callback function or in an IIFE.
```

File: docs/rules/no-shadow.md
Comment (mdjermanovic) [inline docs/rules/no-shadow.md:173]:
```suggestion
Examples of **incorrect** code for the `{ "ignoreOnInitialization": "true" }` option:
```

`{ "ignoreOnInitialization": "true" }` isn't by default.

File: docs/rules/no-shadow.md
Comment (mdjermanovic) [inline docs/rules/no-shadow.md:193]:
```suggestion
The rationale for callback functions is the assumption that they will be called during the initialization, so that at the time when the shadowing variable will be used, the shadowed variable has not yet been initialized.
```

File: docs/rules/no-shadow.md
Comment (mdjermanovic) [inline docs/rules/no-shadow.md:181]:
```suggestion
Because the shadowing variable `x` will shadow the already initialized shadowed variable `x`.
```

File: docs/rules/no-shadow.md
Comment (mdjermanovic) [inline docs/rules/no-shadow.md:190]:
```suggestion
var x = foo(x => x)

var y = (y => y)()
```

To avoid confusion with redeclarations.

**Comment by boutahlilsoufiane** [comment]
Hi @mdjermanovic, thanks a lot for your time and your efforts, thanks a lot for taking the time to explain things more clearly, I think the option `ignoreOnInitialization `is stable due to your great notes.

File: lib/rules/no-shadow.js
Comment (mdjermanovic) [inline lib/rules/no-shadow.js:162]: This would be a false negative:

```js
/* eslint no-shadow: ["error", { ignoreOnInitialization: true }] */

const foo = (x) => {
  bar(x => {})
}
```

The `while(node)` loop finds VariableDeclarator `foo = ...` which has `bar(x => {})` on the right side (in `.init`).

We should either break this loop earlier or check if `outerDef.name` is on the left side.

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

File: lib/rules/no-shadow.js
Comment (boutahlilsoufiane) [reply lib/rules/no-shadow.js]: That happened because I removed the third `if`  from the following snippet, I think we should leave it because it treats the mentioned case:

```
    while (node) {
        if (node.type === "VariableDeclarator") {
            if (isInRange(node.init, location)) {
                return true;
            }
            if (FOR_IN_OF_TYPE.test(node.parent.parent.type) &&
                isInRange(node.parent.parent.right, location)
            ) {
                return true;
            }
            break;
        } else if (node.type === "AssignmentPattern") {
            if (isInRange(node.right, location)) {
                return true;
            }
        } else if (SENTINEL_TYPE.test(node.type)) { // <== I removed this
            break;
        }

        node = node.parent;
    }
```

**Comment by boutahlilsoufiane** [comment]
Hi @mdjermanovic, thanks a lot for your time and your efforts. The comment is addressed.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks! I'll leave this open for a day in case someone else wants to review it before merging.

**Review by nzakas — APPROVED** [review]
Awesome work. Thanks for your patience and responsiveness.

**Comment by mdjermanovic** [comment]
@boutahlilsoufiane great work, thanks for contributing!

**Comment by boutahlilsoufiane** [comment]
Hi @nzakas  and @mdjermanovic  Thanks a lot for approving the changes. Thanks a lot, @mdjermanovic  for your big help, I can't imagine this will be merged without you.

**Pull request merged**
